File: reimplementation/adl2/crates/adl-analysis/src/reconcile.rs
//! Cross/intra-collection reconciliation encoding (Track C keystone).
//!
//! Two analyses (or two object blocks) rarely spell a shared collection the
//! same way — `jets` cut `pt > 30` in one file and `pt > 25` in another are
//! byte-distinct filter DAGs, so structural interning never relates them and
//! their sizes look independent. This module lowers each pure `Base`/`Filtered`
//! chain's membership predicate onto ONE shared generic base element
//! (`base[GENERIC_INDEX]`), so that two predicates over the same base symbol
//! live in the same quantity space and the existing UNSAT-side subset prover
//! can decide refinement.
//!
//! It is PURELY an encoding step: it interns helper quantities and produces a
//! [`Formula`] per candidate collection, but emits no solver fact and reads no
//! verdict. The engine ([`crate::engine`]) consumes the [`ReconEnc`], proves
//! each candidate pair's refinement on the subset side, and only then asserts a
//! derived `size(A) <= size(B)` fact — so nothing here can, by itself, change a
//! verdict. Soundness of the derived facts is argued at the XSUB/XEQ catalog
//! rows and in [`crate::engine::Engine::reconcile`].
//!
//! Fail-closed guards (soundness): a candidate is DROPPED when a filter
//! predicate references a composite binder/reduce element OR a CONCRETE peer
//! element (`Jet[1]`, an angular separation, a collection size) — see
//! [`adl_axioms::encode_elem_pred_generic`]; and a candidate whose base is not
//! a known ext detector object is skipped (private base names collide by
//! spelling). Both were confirmed false-PROVEN paths in adversarial review.
//!
//! DOCUMENTED RESIDUAL (property-alias class, same status as base identity and
//! the float-vs-real gap): distinct source property spellings that canonicalise
//! to ONE key (e.g. `constituents`/`daughters` → `ccountof` in
//! property_vars.txt) ground onto one generic quantity, so a refinement can be
//! proven between them. This is SELF-CONSISTENT — the interpreter and witness
//! realizer read both through the same key, so re-validation agrees — and is a
//! tool-wide modeling choice inherited from the legacy property table, not a
//! reconciliation-specific contradiction.

use adl_axioms::{GENERIC_INDEX, encode_elem_pred_generic};
use adl_formula::{DiagTable, Formula};
use adl_sema::{Collection, ElemPredId, ExtDecls, Hir, Quantity, QuantityId};
use std::collections::BTreeSet;

/// One reconciliation candidate: two distinct filtered collections over the
/// same base, with their membership predicates lowered onto the shared generic
/// element. `phi_a`/`phi_b` are three-valued (opaque leaves are `Unknown`,
/// never dropped) so the engine can honour the FULL superset predicate on the
/// `.under()` side.
pub(crate) struct ReconCandidate {
    pub size_a: QuantityId,
    pub size_b: QuantityId,
    pub phi_a: Formula,
    pub phi_b: Formula,
}

/// The reconciliation encoding for one (merged) unit.
pub(crate) struct ReconEnc {
    pub candidates: Vec<ReconCandidate>,
}

impl ReconEnc {
    /// Every helper quantity the candidate formulas mention (generic-element
    /// properties) plus the two collection sizes per candidate — the engine
    /// must declare these to the solver before asserting anything.
    pub(crate) fn quantities(&self) -> BTreeSet<QuantityId> {
        let mut out = BTreeSet::new();
        for c in &self.candidates {
            out.insert(c.size_a);
            out.insert(c.size_b);
            crate::encode::formula_quantities(&c.phi_a, &mut out);
            crate::encode::formula_quantities(&c.phi_b, &mut out);
        }
        out
    }

    pub(crate) fn is_empty(&self) -> bool {
        self.candidates.is_empty()
    }
}

/// Build the reconciliation encoding: for every candidate pair, lower both
/// filter chains' predicates onto the shared base's generic element. A pair is
/// dropped (fail-closed, NO relation) when a predicate references a residual
/// composite/reduce binder that cannot ground onto one element.
///
/// Mutates `hir.table` (interns the shared base collection, the generic
/// element quantities, and the two collection sizes). MUST run after
/// `emit_axioms` so these helper quantities receive no base axioms of their own
/// — they exist only inside the transient subset frames the engine pushes.
pub(crate) fn build(hir: &mut Hir, ext: &ExtDecls) -> ReconEnc {
    let cands = hir.table.reconciliation_candidates();
    let mut candidates = Vec::new();
    for (a, b) in cands {
        // Both flatten to the same base symbol (guaranteed by the candidate
        // enumeration); re-read to obtain the predicate lists.
        let Some((base_sym, preds_a)) = hir.table.filter_chain(a) else {
            continue;
        };
        let Some((_, preds_b)) = hir.table.filter_chain(b) else {
            continue;
        };
        // Only reconcile over a genuine DETECTOR base (a known ext object:
        // Jet/Electron/Muon/MET/...). An UNKNOWN base name is interned
        // verbatim as a PRIVATE base (resolve_base_name), so two analyses can
        // reuse the same spelling for physically different inputs and collide
        // to one base Symbol — the "same base name = same input" residual is
        // safe only for real shared detector objects, not arbitrary names.
        if ext.base_collection(hir.symbols.display(base_sym)).is_none() {
            continue;
        }
        let base = hir.table.intern_collection(Collection::Base(base_sym));
        let Some(phi_a) = lower(hir, &preds_a, base) else {
            continue;
        };
        let Some(phi_b) = lower(hir, &preds_b, base) else {
            continue;
        };
        let size_a = hir.table.intern_quantity(Quantity::Size(a));
        let size_b = hir.table.intern_quantity(Quantity::Size(b));
        candidates.push(ReconCandidate {
            size_a,
            size_b,
            phi_a,
            phi_b,
        });
    }
    ReconEnc { candidates }
}

/// Conjoin a filter chain's predicates, each lowered onto `base[GENERIC_INDEX]`.
/// Returns `None` if any predicate fails to ground (residual binder/reduce) —
/// the whole pair is then excluded.
fn lower(hir: &mut Hir, preds: &[ElemPredId], base: adl_sema::CollectionId) -> Option<Formula> {
    // Reconciliation diagnostics are never surfaced (opaque leaves resolve to
    // true/false under .over()/.under()), so a throwaway table is correct.
    let mut diags = DiagTable::default();
    let mut parts = Vec::with_capacity(preds.len());
    for &pid in preds {
        let node = hir.elem_pred(pid).node.clone();
        let f = encode_elem_pred_generic(&mut hir.table, &node, base, GENERIC_INDEX, &mut diags)?;
        parts.push(f);
    }
    Some(match parts.len() {
        // A bare filter with no lowered conjuncts keeps every base element.
        0 => Formula::True,
        1 => parts.pop().expect("len == 1"),
        _ => Formula::And(parts),
    })
}

File: cpp/libs/analysis/src/reconcile.cpp
#include "adl2/analysis/reconcile.hpp"

#include "adl2/analysis/encode.hpp"
#include "adl2/axioms/axioms.hpp"
#include "adl2/sema/quantity.hpp"

#include <optional>
#include <utility>

namespace adl2::analysis {
namespace {

using adl2::axioms::GENERIC_INDEX;
using adl2::axioms::encode_elem_pred_generic;
using adl2::formula::DiagTable;
using adl2::formula::Formula;
using adl2::sema::Collection;
using adl2::sema::CollectionId;
using adl2::sema::CollectionKind;
using adl2::sema::ElemPredId;
using adl2::sema::ExtDecls;
using adl2::sema::Hir;
using adl2::sema::Quantity;
using adl2::sema::QuantityId;
using adl2::sema::Symbol;

std::optional<Formula> lower(Hir& hir, const std::vector<ElemPredId>& preds,
                             CollectionId base) {
  DiagTable diags;
  std::vector<Formula> parts;
  parts.reserve(preds.size());
  for (auto pid : preds) {
    const auto& node = hir.elem_pred(pid).node;
    auto f = encode_elem_pred_generic(hir.table, node, base, GENERIC_INDEX, diags);
    if (!f) return std::nullopt;
    parts.push_back(std::move(*f));
  }
  if (parts.empty()) return Formula::ttrue();
  if (parts.size() == 1) return std::move(parts.front());
  return Formula::of_and(std::move(parts));
}

std::vector<ReconNearMiss> near_misses(Hir& hir, const ExtDecls& ext) {
  struct Chain {
    CollectionId id;
    Symbol base;
    std::vector<ElemPredId> preds;
  };
  std::vector<Chain> chains;
  for (std::size_t i = 0; i < hir.table.collections().size(); ++i) {
    CollectionId id{static_cast<std::uint32_t>(i)};
    if (hir.table.collection(id).kind != CollectionKind::Filtered) continue;
    Symbol base;
    std::vector<ElemPredId> preds;
    if (hir.table.filter_chain(id, base, preds)) {
      chains.push_back(Chain{id, base, std::move(preds)});
    }
  }
  std::vector<ReconNearMiss> out;
  for (std::size_t i = 0; i < chains.size(); ++i) {
    for (std::size_t j = i + 1; j < chains.size(); ++j) {
      if (chains[i].base == chains[j].base) continue;
      std::string name_a = hir.symbols.display(chains[i].base);
      std::string name_b = hir.symbols.display(chains[j].base);
      bool known_a = ext.base_collection(name_a) != nullptr;
      bool known_b = ext.base_collection(name_b) != nullptr;
      if (known_a && known_b) continue;
      CollectionId probe = hir.table.intern_collection(Collection::of_base(chains[i].base));
      auto fa = lower(hir, chains[i].preds, probe);
      auto fb = lower(hir, chains[j].preds, probe);
      if (!fa || !fb) continue;
      if (*fa == *fb) {
        ReconNearMiss n;
        n.coll_a = chains[i].id;
        n.coll_b = chains[j].id;
        n.base_a = std::move(name_a);
        n.base_b = std::move(name_b);
        out.push_back(std::move(n));
      }
    }
  }
  return out;
}

}  // namespace

std::set<QuantityId> ReconEnc::quantities() const {
  std::set<QuantityId> out;
  for (const auto& c : candidates) {
    out.insert(c.size_a);
    out.insert(c.size_b);
    formula_quantities(c.phi_a, out);
    formula_quantities(c.phi_b, out);
  }
  return out;
}

ReconEnc build_recon(Hir& hir, const ExtDecls& ext) {
  auto cands = hir.table.reconciliation_candidates();
  ReconEnc enc;
  for (const auto& pair : cands) {
    CollectionId a = pair.first;
    CollectionId b = pair.second;
    Symbol base_sym;
    std::vector<ElemPredId> preds_a;
    std::vector<ElemPredId> preds_b;
    if (!hir.table.filter_chain(a, base_sym, preds_a)) continue;
    Symbol ignore;
    if (!hir.table.filter_chain(b, ignore, preds_b)) continue;
    if (ext.base_collection(hir.symbols.display(base_sym)) == nullptr) {
      ReconSkip s;
      s.coll_a = a;
      s.coll_b = b;
      s.reason =
          "base `" + std::string(hir.symbols.display(base_sym)) +
          "` is not a known detector object, so a shared spelling "
          "cannot be assumed to mean a shared input";
      enc.skipped.push_back(std::move(s));
      continue;
    }
    CollectionId base = hir.table.intern_collection(Collection::of_base(base_sym));
    auto phi_a = lower(hir, preds_a, base);
    auto phi_b = lower(hir, preds_b, base);
    if (!phi_a || !phi_b) {
      ReconSkip s;
      s.coll_a = a;
      s.coll_b = b;
      s.reason =
          "a cut references a composite/peer element that cannot ground "
          "onto one shared element";
      enc.skipped.push_back(std::move(s));
      continue;
    }
    ReconCandidate c;
    c.size_a = hir.table.intern_quantity(Quantity::size(a));
    c.size_b = hir.table.intern_quantity(Quantity::size(b));
    c.phi_a = std::move(*phi_a);
    c.phi_b = std::move(*phi_b);
    c.coll_a = a;
    c.coll_b = b;
    enc.candidates.push_back(std::move(c));
  }
  enc.near_misses = near_misses(hir, ext);
  return enc;
}

}  // namespace adl2::analysis
